### PR TITLE
FileJournal: Support batch peak and pop from writeq

### PR DESCRIPTION
**Pull request by XinzeChi**
This would reduce writeq_lock contentions a lot.
Signed-off-by: Xinze Chi <xinze@xsky.com>

### CODE REVIEW

**Comment by XinzeChi** [comment]
@liewegas , is it make sense? It perform better in ssd.

**Comment by XinzeChi** [comment]
@jack-changtao , There are too many places where we use writeq. If do as you said, I think this would be a little complicated. @liewegas

**Comment by yuyuyu101** [comment]
we need a benchmark result refer to reviewers.

**Comment by XinzeChi** [comment]
image

Above is the benchmark based on https://github.com/XinzeChi/ceph/commit/cde544ed4e7fefa0ba80a5ff67794bc2c475e3b8.

The hardware environment is:
CPU: Intel(R) Xeon(R) CPU E5-2620 v2 @ 2.10GHz
SSD: Intel P3710

**Comment by XinzeChi** [comment]
image

Above is the benchmark based on https://github.com/XinzeChi/ceph/commit/cde544ed4e7fefa0ba80a5ff67794bc2c475e3b8.
The hardware environment is:
CPU: Intel(R) Xeon(R) CPU E5-2620 v2 @ 2.10GHz
SSD: Intel P3710

File: src/os/FileJournal.cc [hunk 1]
```diff
@@ -1666,6 +1675,20 @@ void FileJournal::pop_write()
...
+  writeq.swap(items);
+}
+
+void FileJournal::backfill_write(list<write_item> &items)
+{
+  assert(write_lock.is_locked());
+  Mutex::Locker locker(writeq_lock);
+  writeq.splice(writeq.begin(), items);
+}
+
```
Comment (liewegas) [inline src/os/FileJournal.cc]: I'd call these batch_shift_writeq() and batch_unshift_writeq()?  (shift/unshift being the opposite of pop/push, as in perl).

File: src/os/FileJournal.cc [hunk 2]
```diff
@@ -1666,6 +1675,20 @@ void FileJournal::pop_write()
   writeq.pop_front();
 }
 
+void FileJournal::batch_peek_write(list<write_item> &items)
+{
+  assert(write_lock.is_locked());
+  Mutex::Locker locker(writeq_lock);
+  writeq.swap(items);
+}
```
Comment (liewegas) [inline src/os/FileJournal.cc]: batch_pop_write (it's not peek, since we're removing the items)

File: src/os/FileJournal.cc [hunk 3]
```diff
@@ -1666,6 +1675,20 @@ void FileJournal::pop_write()
...
+
+void FileJournal::backfill_write(list<write_item> &items)
+{
+  assert(write_lock.is_locked());
+  Mutex::Locker locker(writeq_lock);
+  writeq.splice(writeq.begin(), items);
+}
```
Comment (liewegas) [inline src/os/FileJournal.cc]: batch_unpop_write()

**Comment by liewegas** [comment]
Hmm, was that test looking just at latency for the journal or from the client?

**Comment by XinzeChi** [comment]
the test is base on https://github.com/XinzeChi/ceph/commit/cde544ed4e7fefa0ba80a5ff67794bc2c475e3b8. I think https://github.com/XinzeChi/ceph/commit/cde544ed4e7fefa0ba80a5ff67794bc2c475e3b8 is good benckmark tool for filejournal?

**Comment by liewegas** [comment]
Ah. My concern if this buys us only a few % for just the journal and the journal is only a fraction of the overall write latency it may not be worth the risk of breaking journal replay or compatibility with other versions.  Note that the new BlueStore backend won't use any of this code, so Jewel is likely the last release where FileStore will be the default.

**Comment by XinzeChi** [comment]
I think you mention https://github.com/ceph/ceph/pull/6856 instead of this patch. Tthis patch is clean and safe.

**Comment by liewegas** [comment]
Ah, yes, mixed them up.  This is simple and clean.

**Comment by liewegas** [comment]
But, please do keep in mind that this code isn't going to be used for too much longer.  You might want to focus efforts on optimizations with the new backend(s), or in the layers above ObjectStore.

**Pull request merged**